src/librustc_mir/borrow_check/nll/region_infer/error_reporting/outlives_suggestion.rs

```diff
@@ -0,0 +1,428 @@
+            Level::Help,
+            "the following suggestions may resolve your lifetime errors",
+        );
+
+        // Build the diagnostic
+        let mut added_names = self.suggestion_name_implicit_lifetimes(rinfcx, infcx, &mut diag);
+        let mut added_constraints = self.suggestion_outlives_constraints(rinfcx, infcx, &mut diag);
+
+        // Buffer the diagnostic or cancel it if there was no suggestion.
+        if added_names || added_constraints {
+            diag.buffer(errors_buffer);
+        } else {
+            diag.cancel();
+        }
+    }
+
+    /// Add a suggestion to give names to any implicit lifetimes. This may require adding generics
+    /// (i.e. `<'a>`) if all of the lifetimes are implicit. It also requires making lifetimes
+    /// explicit in names.
+    fn suggestion_name_implicit_lifetimes<'tcx>(
+        &self,
+        rinfcx: &RegionInferenceContext<'tcx>,
+        infcx: &InferCtxt<'_, '_, 'tcx>,
+        diag: &mut DiagnosticBuilder<'_>,
+    ) -> bool {
+        // No lifetimes to name... nothing to do...
+        if self.names_to_add.is_empty() {
+            debug!("No names to add");
+            return false;
+        }
+
+        // If needed add a list of named generics.
+        let tcx = infcx.tcx;
+
+        let error_region = rinfcx
+            .to_error_region(
+                self.constraints_to_add
+                    .iter()
+                    .next()
+                    .map(|(&fr, _)| fr)
+                    .unwrap(),
+            )
+            .expect("unable to find error region");
+        let fn_def_id = error_region.free_region_binding_scope(tcx); // FIXME: this could be an impl...
+        let fn_node_id = tcx
+            .hir
+            .as_local_node_id(fn_def_id)
+            .expect("unable to get node id");
+        let generics = tcx.hir.get_generics(fn_def_id).unwrap();
+
+        let mut generics_sug = String::new();
+
+        generics_sug += "<";
+
+        let generics_span = if let Some(generics_span) = tcx.hir.get_generics_span(fn_def_id) {
+            for (i, generic) in generics.params.iter().enumerate() {
+                if i > 0 {
+                    generics_sug += ", ";
+                }
+
+                generics_sug += &tcx
+                    .sess
+                    .source_map()
+                    .span_to_snippet(generic.span)
+                    .expect("unable to get snippet");
+            }
+
+            generics_span
+        } else {
+            // Could be a method or a fn
+            match tcx.hir.get(fn_node_id) {
+                Node::Item(
+                    item @ Item {
+                        node: ItemKind::Fn(..),
+                        ..
+                    },
+                ) => {
+                    // FIXME: HACK! Assume the name and the `fn` keyword are separated by a single
+                    // space. yeesh...
+                    let item_span = item.span;
```

Comment:
@estebank I want to get the name of the function here. This code is rather hacky. I feel like there must be a better way than this. Do you have any ideas?